cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/replication/PullDocIdsFilter.java

```diff
@@ -0,0 +1,40 @@
+package com.cloudant.sync.replication;
+
+import com.cloudant.sync.internal.util.JSONUtils;
+import com.cloudant.sync.internal.util.Misc;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * <p>Sets the list of doc IDs to use as filtering criteria when a pull replication calls the
+ * source database's {@code _changes} feed.
+ * </p>
+ * Note: Doc IDs filtering are supported only when replicating against a CouchDB 2.x compliant database
+ *
+ * @see
+ * <a target="_blank" href="https://console.bluemix.net/docs/services/Cloudant/api/database.html#get-changes">See doc_ids filtering</a>
+ */
+public class PullDocIdsFilter extends PullFilter {
+
+    private static final String DOC_IDS_FILTER_NAME = "_doc_ids";
+
+    public PullDocIdsFilter(List<String> docIds) {
+        super(DOC_IDS_FILTER_NAME);
+
+        Misc.checkState(docIds != null && !docIds.isEmpty(),
+                "Doc Ids should not be empty in a Doc Ids Filter");
+
+        List<String> internalDocIds = new ArrayList<String>(docIds);
+        Collections.sort(internalDocIds);
+
+        Map<String, Object> parameters = new HashMap<String, Object>();
+        parameters.put("doc_ids", internalDocIds);
+
+        setParameters(parameters);
```

Comment:
OK I can see the value in sorting the doc ids, but there is no need to call the `setParameters` helper. You are creating a `HashMap` temporarily here and then throwing it away when you call `setParameters`. Instead just do

```java
Map<String, Object> internalParameters = new HashMap<String, Object>();
internalParameters.put("doc_ids", internalDocIds);
parameters = Collections.unmodifiableMap(internalParams);
```

Reply:
Note that you will need to make `name` and `parameters` protected in `PullFilter` for the above and for my comment relating to setting `name` directly in the `PullDocIdsFilter` - I think it's a worthwhile compromise to get the `final` correctness right.

Reply:
It is not clear for me if you want to keep the ```final``` modifier for ```name``` and ```parameters```. In this case we can not set the value for these members from the constructor of a subclass.

Reply:
OK, yes I can see why we can't keep parameters `final`